### DISCUSSION TITLE
Expanding the `ensure()` pattern

### DISCUSSION THREAD

**Discussion by mattyg** [holochain/holochain#2409] [General]
I really like the `ensure()` function and pattern available to us for Paths. I feel like it would be helpful to make this available for entries and links as well. 

There are many of scenarios where you might only *want* a single action / link to be published, but it's not possible to enforce this, so the ensure pattern can be used as a soft way to reduce the amount of duplication in the dht. 

One example I've run into is an Decision entry that represents the outcome of a governance vote. Lets say the entry includes the actionhashs of all tallied Vote entries, plus the voting rules, plus the actionhash of the current time from an "official" timekeeper agent.

Anyone should be able to publish this decision, because we don't want some subset of agents to be able to delay / stop a governance decision from proceeding. However, we only *need* a single Decision to be published.

**Comment by maackle** [comment]
Ensuring links would require that links be constructable and hashable, so they can be looked up by hash. This seems like no major problem but I'm making note of it as a dependency.
